# Forecasting English Premier League Results

In the previous notebook, we saw how to fit a simple model to account for the variation in game results. Our simple model explained the variation of results as a function of the ratio of TM values for the two teams, and home advantage. We found that this model came close to approximating the reliability of bookmakers in predicting results. 

Now we want to test the reliability of our model as a *forecasting* model. The difference from a statistical point of view is this: with the previous model we assessed the ability of the model to predict results "within sample" - meaning that the same data used to estimate the regression model was used to test the accuracy of the model, while now we want to test the accuracy of the model "out-of-sample" - to test the reliability of the model that has been estimated using one dataset, when predicting the results based on another dataset.

Forecasting events that have not yet happened is by definition out-of-sample forecasting. We are going to use the Premier League data for the season 2019/20. Due to Covid-19, the Premier League season was suspended on March 13, 2020, leaving 89 of the 380 games unplayed. At the time of this writing (April 8, 2020), it was not clear if it will ever be possible to complete the season, although the League still holds out some hope that it might be. 
We will take the following steps:

1. Use data on the first 198 games played (all games played in 2019) to estimate our regression model
2. Use the coefficients from this model together with the TM values for the clubs to forecast the outcome of 90 played in 2020
3. We will compare this out-of-sample forecasting performance to the betting odds and also to the weekly forecasts of the popular data analytics website, Nate Silver's FiveThirtyEight.

Note that because TM values are typically available in advance of any game being played, you can apply the model described here to real time game forecasting, not just for the Premier League, but for any league which is covered by TransferMarkt.

In [1]:
# This allows us to show the full screen width

from IPython.display import display, HTML

display(HTML(data="""
<style>
    div#notebook-container    { width: 95%; }
    div#menubar-container     { width: 65%; }
    div#maintoolbar-container { width: 99%; }
</style>
"""))

In [2]:
# install the packages we need

import pandas as pd
import numpy as np

In [3]:
# load the data for the 2019/20 season. This includes 380 scheduled games, of which only 288 had been played before the suspension of the season on March 13, 2020

season19_20 = pd.read_excel("../../Data/Week 3/EPL19-20.xlsx")
season19_20

date          Home team                 away team  \
0    2019-10-08 00:00:00    AFC Bournemouth          Sheffield United   
1    2019-10-08 00:00:00            Burnley               Southampton   
2    2019-10-08 00:00:00     Crystal Palace                   Everton   
3    2019-10-08 00:00:00  Tottenham Hotspur               Aston Villa   
4    2019-10-08 00:00:00            Watford  Brighton and Hove Albion   
..                   ...                ...                       ...   
375  2020-05-17 00:00:00     Leicester City         Manchester United   
376  2020-05-17 00:00:00    Manchester City              Norwich City   
377  2020-05-17 00:00:00   Newcastle United                 Liverpool   
378  2020-05-17 00:00:00        Southampton          Sheffield United   
379  2020-05-17 00:00:00    West Ham United               Aston Villa   

     notplayed  month   day    year  FTHG  FTAG  FTR  B365H  B365D  B365A  \
0            0    8.0  10.0  2019.0   1.0   1.0    D   1.95   3.60   3.60   
1            0    8.0  10.0  2019.0   3.0   0.0    H   2.62   3.20   2.75   
2            0    8.0  10.0  2019.0   0.0   0.0    D   3.00   3.25   2.37   
3            0    8.0  10.0  2019.0   3.0   1.0    H   1.30   5.25  10.00   
4            0    8.0  10.0  2019.0   0.0   3.0    A   1.90   3.40   4.00   
..         ...    ...   ...     ...   ...   ...  ...    ...    ...    ...   
375          1    NaN   NaN     NaN   NaN   NaN  NaN    NaN    NaN    NaN   
376          1    NaN   NaN     NaN   NaN   NaN  NaN    NaN    NaN    NaN   
377          1    NaN   NaN     NaN   NaN   NaN  NaN    NaN    NaN    NaN   
378          1    NaN   NaN     NaN   NaN   NaN  NaN    NaN    NaN    NaN   
379          1    NaN   NaN     NaN   NaN   NaN  NaN    NaN    NaN    NaN   

      TMhome  TMaway  538hpr  538dpr  538apr  
0     281.70   62.33    0.55    0.21    0.24  
1     180.68  209.70    0.45    0.30    0.26  
2     207.50  457.20    0.36    0.38    0.26  
3     881.55  140.40    0.73    0.09    0.18  
4     214.52  180.99    0.51    0.24    0.25  
..       ...     ...     ...     ...     ...  
375   343.13  644.63     NaN     NaN     NaN  
376  1140.00   81.54     NaN     NaN     NaN  
377   225.97  959.18     NaN     NaN     NaN  
378   209.70   62.33     NaN     NaN     NaN  
379   299.03  140.40     NaN     NaN     NaN  

[380 rows x 18 columns]

In [4]:
season19_20.describe()

notplayed       month         day         year        FTHG  \
count  380.000000  288.000000  288.000000   288.000000  288.000000   
mean     0.242105    7.614583   16.027778  2019.312500    1.506944   
std      0.428922    4.193166    9.400085     0.464319    1.192290   
min      0.000000    1.000000    1.000000  2019.000000    0.000000   
25%      0.000000    2.000000    8.000000  2019.000000    1.000000   
50%      0.000000    9.000000   17.500000  2019.000000    1.000000   
75%      0.000000   11.000000   24.000000  2020.000000    2.000000   
max      1.000000   12.000000   31.000000  2020.000000    8.000000   

             FTAG       B365H       B365D       B365A       TMhome  \
count  288.000000  288.000000  288.000000  288.000000   380.000000   
mean     1.215278    2.867257    4.319618    4.774410   402.745500   
std      1.207775    2.277826    1.514852    4.165434   303.061162   
min      0.000000    1.070000    3.100000    1.140000    62.330000   
25%      0.000000    1.607500    3.400000    2.300000   200.872500   
50%      1.000000    2.200000    3.750000    3.350000   279.340000   
75%      2.000000    3.100000    4.750000    5.750000   588.942500   
max      9.000000   15.000000   13.000000   26.000000  1140.000000   

            TMaway      538hpr      538dpr      538apr  
count   380.000000  288.000000  288.000000  288.000000  
mean    402.745500    0.464549    0.268993    0.266771  
std     303.061162    0.190143    0.134825    0.131910  
min      62.330000    0.070000    0.010000    0.030000  
25%     200.872500    0.347500    0.200000    0.200000  
50%     279.340000    0.450000    0.250000    0.250000  
75%     588.942500    0.582500    0.290000    0.290000  
max    1140.000000    0.920000    0.790000    0.750000

We now create a variable 'B365res' which is the predicted result judged by the outcome with the highest probability as implied by the B365 betting odds. The odds here are decimal odds, which means that the lowest value is the result with the highest probability based on the odds. 

Note that we identify the outcome using inequalities, e.g. if the decimal odds of a home win are smaller than the odds of a draw, and the home win-win odds are smaller than the odds of an away win, then we say that the predicted outcome is a home win. But what if the odds of a home win were smaller than the odds of a draw, but were equal to the odds of an away win? Our code does not give any indication about how to resolve that tie, and therefore our formula would return an empty cell for that row. In our data, there are no examples of this, and generally this is rare, but if we came across that we would have to decide what to do. If there were many cases we would have to create a separate category of outcome for assessing the reliability of the odds. If there were a small number of cases we could randomly resolve the tie in favour of one outcome or the other. This problem should never arise when it comes to our regression model, since the probabilities will always be estimated to many decimal places, and the chances of two identical outcomes is almost zero. 

In [5]:
season19_20['B365res']= np.where((season19_20['B365H']<season19_20['B365D']) & (season19_20['B365H']<season19_20['B365A']),'H',\
                            np.where((season19_20['B365D']<season19_20['B365H']) & (season19_20['B365D']<season19_20['B365A']),'D',\
                                    np.where((season19_20['B365A']<season19_20['B365H']) & (season19_20['B365A']<season19_20['B365D']),'A',"")))
pd.set_option('display.max_rows', 400)
season19_20

date                 Home team                 away team  \
0    2019-10-08 00:00:00           AFC Bournemouth          Sheffield United   
1    2019-10-08 00:00:00                   Burnley               Southampton   
2    2019-10-08 00:00:00            Crystal Palace                   Everton   
3    2019-10-08 00:00:00         Tottenham Hotspur               Aston Villa   
4    2019-10-08 00:00:00                   Watford  Brighton and Hove Albion   
5    2019-10-08 00:00:00           West Ham United           Manchester City   
6    2019-11-08 00:00:00            Leicester City   Wolverhampton Wanderers   
7    2019-11-08 00:00:00         Manchester United                   Chelsea   
8    2019-11-08 00:00:00          Newcastle United                   Arsenal   
9             17/08/2019                   Arsenal                   Burnley   
10            17/08/2019               Aston Villa           AFC Bournemouth   
11            17/08/2019  Brighton and Hove Albion           West Ham United   
12            17/08/2019                   Everton                   Watford   
13            17/08/2019           Manchester City         Tottenham Hotspur   
14            17/08/2019              Norwich City          Newcastle United   
15            17/08/2019               Southampton                 Liverpool   
16            18/08/2019                   Chelsea            Leicester City   
17            18/08/2019          Sheffield United            Crystal Palace   
18            19/08/2019   Wolverhampton Wanderers         Manchester United   
19            23/08/2019               Aston Villa                   Everton   
20            24/08/2019  Brighton and Hove Albion               Southampton   
21            24/08/2019                 Liverpool                   Arsenal   
22            24/08/2019         Manchester United            Crystal Palace   
23            24/08/2019              Norwich City                   Chelsea   
24            24/08/2019          Sheffield United            Leicester City   
25            24/08/2019                   Watford           West Ham United   
26            25/08/2019           AFC Bournemouth           Manchester City   
27            25/08/2019         Tottenham Hotspur          Newcastle United   
28            25/08/2019   Wolverhampton Wanderers                   Burnley   
29            31/08/2019                   Burnley                 Liverpool   
30            31/08/2019                   Chelsea          Sheffield United   
31            31/08/2019            Crystal Palace               Aston Villa   
32            31/08/2019            Leicester City           AFC Bournemouth   
33            31/08/2019           Manchester City  Brighton and Hove Albion   
34            31/08/2019          Newcastle United                   Watford   
35            31/08/2019               Southampton         Manchester United   
36            31/08/2019           West Ham United              Norwich City   
37   2019-01-09 00:00:00                   Arsenal         Tottenham Hotspur   
38   2019-01-09 00:00:00                   Everton   Wolverhampton Wanderers   
39            14/09/2019  Brighton and Hove Albion                   Burnley   
40            14/09/2019                 Liverpool          Newcastle United   
41            14/09/2019         Manchester United            Leicester City   
42            14/09/2019              Norwich City           Manchester City   
43            14/09/2019          Sheffield United               Southampton   
44            14/09/2019         Tottenham Hotspur            Crystal Palace   
45            14/09/2019   Wolverhampton Wanderers                   Chelsea   
46            15/09/2019           AFC Bournemouth                   Everton   
47            15/09/2019                   Watford                   Arsenal   
48            16/09/2019               Aston Villa           West Ham United   
49            20/09/

One way to check the reliability of the betting odds is to use pd.crosstab() to show the count's actual results against the predicted results. We show this below.

What jumps out at you from this crosstab is that the bookmaker odds *never* predict a draw! Later on, you will find this is true of our model, and also true of the forecasting model published by Nate Silver's FiveThirtyEight. But draws are very common in soccer - in any league they typically account for around a quarter of all results! We will discuss this finding again when we look at the forecasts in more detail below. 

In [6]:
pd.crosstab(season19_20['FTR'], season19_20['B365res'],dropna= True)

B365res   A    H
FTR             
A        49   38
D        20   52
H        26  103

## Estimating our regression model

Though not all of the games for the season have been played, we have TM values for all the teams. These are the values published just before the season started. TransferMarkt updates its TM values for the Premier League several times a year, and it might be possible to improve the accuracy of forecasts by always choosing the most recent TM values, but for the purposes of showing how to apply the forecast model, the values from the beginning of the season will be good enough.

The TM values for the home team and away team are loaded into the df, so we only need to create the log of the ratio of home team TM value to away team TM value.

In [7]:
# create the log of TM ratios 
season19_20['lhTMratio'] = np.log(season19_20['TMhome']/season19_20['TMaway'])

Now we create a numerical value for the actual outcome of the game (FTR) where H = 2, D = 1 and A = 0.

In [8]:
# create a value = 2 if homewin, 1 if draw, 0 if awaywin
season19_20['winvalue'] = np.where(season19_20['FTR'] == 'H', 2 ,(np.where(season19_20['FTR'] == 'D', 1, 0)))
season19_20

date                 Home team                 away team  \
0    2019-10-08 00:00:00           AFC Bournemouth          Sheffield United   
1    2019-10-08 00:00:00                   Burnley               Southampton   
2    2019-10-08 00:00:00            Crystal Palace                   Everton   
3    2019-10-08 00:00:00         Tottenham Hotspur               Aston Villa   
4    2019-10-08 00:00:00                   Watford  Brighton and Hove Albion   
5    2019-10-08 00:00:00           West Ham United           Manchester City   
6    2019-11-08 00:00:00            Leicester City   Wolverhampton Wanderers   
7    2019-11-08 00:00:00         Manchester United                   Chelsea   
8    2019-11-08 00:00:00          Newcastle United                   Arsenal   
9             17/08/2019                   Arsenal                   Burnley   
10            17/08/2019               Aston Villa           AFC Bournemouth   
11            17/08/2019  Brighton and Hove Albion           West Ham United   
12            17/08/2019                   Everton                   Watford   
13            17/08/2019           Manchester City         Tottenham Hotspur   
14            17/08/2019              Norwich City          Newcastle United   
15            17/08/2019               Southampton                 Liverpool   
16            18/08/2019                   Chelsea            Leicester City   
17            18/08/2019          Sheffield United            Crystal Palace   
18            19/08/2019   Wolverhampton Wanderers         Manchester United   
19            23/08/2019               Aston Villa                   Everton   
20            24/08/2019  Brighton and Hove Albion               Southampton   
21            24/08/2019                 Liverpool                   Arsenal   
22            24/08/2019         Manchester United            Crystal Palace   
23            24/08/2019              Norwich City                   Chelsea   
24            24/08/2019          Sheffield United            Leicester City   
25            24/08/2019                   Watford           West Ham United   
26            25/08/2019           AFC Bournemouth           Manchester City   
27            25/08/2019         Tottenham Hotspur          Newcastle United   
28            25/08/2019   Wolverhampton Wanderers                   Burnley   
29            31/08/2019                   Burnley                 Liverpool   
30            31/08/2019                   Chelsea          Sheffield United   
31            31/08/2019            Crystal Palace               Aston Villa   
32            31/08/2019            Leicester City           AFC Bournemouth   
33            31/08/2019           Manchester City  Brighton and Hove Albion   
34            31/08/2019          Newcastle United                   Watford   
35            31/08/2019               Southampton         Manchester United   
36            31/08/2019           West Ham United              Norwich City   
37   2019-01-09 00:00:00                   Arsenal         Tottenham Hotspur   
38   2019-01-09 00:00:00                   Everton   Wolverhampton Wanderers   
39            14/09/2019  Brighton and Hove Albion                   Burnley   
40            14/09/2019                 Liverpool          Newcastle United   
41            14/09/2019         Manchester United            Leicester City   
42            14/09/2019              Norwich City           Manchester City   
43            14/09/2019          Sheffield United               Southampton   
44            14/09/2019         Tottenham Hotspur            Crystal Palace   
45            14/09/2019   Wolverhampton Wanderers                   Chelsea   
46            15/09/2019           AFC Bournemouth                   Everton   
47            15/09/2019                   Watford                   Arsenal   
48            16/09/2019               Aston Villa           West Ham United   
49            20/09/

We are going to use the games from the season played in the calendar year of 2019 to estimate our regression model (198 games), and then use the estimated coefficients from that model to forecast the games played in 2020 (90 games). The split here is arbitrary - we could have chosen the balance differently. We should expect the regression model to improve as we increase the number of games included, but also to deteriorate as the time elapsed between the games used for estimation and the games forecast increases (since the earlier information becomes out of date). Viewed as a forecasting model, however, the point is that from the end of December 2019 it would have been possible to use the model to forecast the outcome of games played in 2020, *before* the games had actually been played.

We first define our subset of games for estimation:

In [9]:
season19 = season19_20[:198]
season19

date                 Home team                 away team  \
0    2019-10-08 00:00:00           AFC Bournemouth          Sheffield United   
1    2019-10-08 00:00:00                   Burnley               Southampton   
2    2019-10-08 00:00:00            Crystal Palace                   Everton   
3    2019-10-08 00:00:00         Tottenham Hotspur               Aston Villa   
4    2019-10-08 00:00:00                   Watford  Brighton and Hove Albion   
5    2019-10-08 00:00:00           West Ham United           Manchester City   
6    2019-11-08 00:00:00            Leicester City   Wolverhampton Wanderers   
7    2019-11-08 00:00:00         Manchester United                   Chelsea   
8    2019-11-08 00:00:00          Newcastle United                   Arsenal   
9             17/08/2019                   Arsenal                   Burnley   
10            17/08/2019               Aston Villa           AFC Bournemouth   
11            17/08/2019  Brighton and Hove Albion           West Ham United   
12            17/08/2019                   Everton                   Watford   
13            17/08/2019           Manchester City         Tottenham Hotspur   
14            17/08/2019              Norwich City          Newcastle United   
15            17/08/2019               Southampton                 Liverpool   
16            18/08/2019                   Chelsea            Leicester City   
17            18/08/2019          Sheffield United            Crystal Palace   
18            19/08/2019   Wolverhampton Wanderers         Manchester United   
19            23/08/2019               Aston Villa                   Everton   
20            24/08/2019  Brighton and Hove Albion               Southampton   
21            24/08/2019                 Liverpool                   Arsenal   
22            24/08/2019         Manchester United            Crystal Palace   
23            24/08/2019              Norwich City                   Chelsea   
24            24/08/2019          Sheffield United            Leicester City   
25            24/08/2019                   Watford           West Ham United   
26            25/08/2019           AFC Bournemouth           Manchester City   
27            25/08/2019         Tottenham Hotspur          Newcastle United   
28            25/08/2019   Wolverhampton Wanderers                   Burnley   
29            31/08/2019                   Burnley                 Liverpool   
30            31/08/2019                   Chelsea          Sheffield United   
31            31/08/2019            Crystal Palace               Aston Villa   
32            31/08/2019            Leicester City           AFC Bournemouth   
33            31/08/2019           Manchester City  Brighton and Hove Albion   
34            31/08/2019          Newcastle United                   Watford   
35            31/08/2019               Southampton         Manchester United   
36            31/08/2019           West Ham United              Norwich City   
37   2019-01-09 00:00:00                   Arsenal         Tottenham Hotspur   
38   2019-01-09 00:00:00                   Everton   Wolverhampton Wanderers   
39            14/09/2019  Brighton and Hove Albion                   Burnley   
40            14/09/2019                 Liverpool          Newcastle United   
41            14/09/2019         Manchester United            Leicester City   
42            14/09/2019              Norwich City           Manchester City   
43            14/09/2019          Sheffield United               Southampton   
44            14/09/2019         Tottenham Hotspur            Crystal Palace   
45            14/09/2019   Wolverhampton Wanderers                   Chelsea   
46            15/09/2019           AFC Bournemouth                   Everton   
47            15/09/2019                   Watford                   Arsenal   
48            16/09/2019               Aston Villa           West Ham United   
49            20/09/

Once again we are going to use an ordered logistic regression to estimate the results. First, we import the package for running ordered logit model.

In [10]:
from bevel.linear_ordinal_regression import OrderedLogit
ol = OrderedLogit()

Now  we run the ordered logit regression of game outcome (winvalue) on the TM ratio:

In [11]:
ol.fit(season19['lhTMratio'], season19['winvalue'])
ol.print_summary()

n=198
                  beta  se(beta)      p  lower 0.95  upper 0.95     
attribute names                                                     
column_1        0.5553    0.1298 0.0000      0.3009      0.8096  ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 

Somers' D = 0.252


We can see the value of the coefficient of lhTMratio is 0.555 and the standard error is 0.1298, yielding a t-statistic of over 4 implying that the coefficient is statistically significant at the 0.01 level and better. The higher the ratio, the better the outcome, as viewed by the home team. 

Recall that in our model for the seasons 2010/11 to 2018/19 we found that the coefficient of lhTMratio is 0.528 and the standard error is 0.0323. The coefficient estimate is quite similar, though the precision of the estimate was somewhat larger, which can be attributed to the larger sample size. Indeed, we could use those estimates to forecast all of the 2019/20 games.  In practice, the ideal procedure would be to use coefficients from the previous season to forecast results at the beginning of a new season, and then transfer to estimates based on the current seasons once enough games have been played (e.g. 100 or more). Of course, in each case, the most recent TM values should be used for forecasting.

We want to convert our estimates into probabilities, and to do this we need the estimates of the intercepts as well. the coefficient for lhTMratio is stored as ol.coef_[0], while the threshold between A and D is stored as ol.coef_[1] the threshold between D and H is stored as ol.coef_[2] We can print these out with appropriate names:

In [12]:
#%% To get the coefficients and the intercepts
print(f'beta = {ol.coef_[0]:.4f}')
print(f'interceptAD = {ol.coef_[1]:.4f}')
print(f'interceptDH = {ol.coef_[2]:.4f}')

beta = 0.5553
interceptAD = -0.7678
interceptDH = 0.3216


To generate the forecast probabilities we need to manipulate the coefficients. The logit regression equation has the form log(p/(1-p)) = a + bX. By rearranging this equation we can obtain p = 1/(1+ exp(a +bX)).

For each game, we know X (lhTMratio) and now we know b (0.555). Since this is an ordered logit with three possible outcomes, the probability of the worst outcome A (viewed by the home team) depends on intercept_AD (ol.coef_[1]), the probability of the middle outcome D depends on intercept_AD and intercept_DH (ol.coef_[1] and ol.coef_[2]), while the probability of the best outcome depends on intercept_DH.

If we calculate the probability of A first, using ol.coef_[0] and ol.coef_[1], then when calculating the probability of D we can use the fact that we already have the probability of A, and also when calculating the probability of H we can use the fact that we already have the probability of D.

Thus, we now create the predicted values of the H, A and D probabilities from our model. We can create a prediction for every game in the season, so we now apply the formulas to the season19_20 df, not the season19 df that we used to generate the regression estimates:

In [13]:
# Predicted probabilities

season19_20['predA'] = 1/(1+np.exp(-(ol.coef_[1]-ol.coef_[0]*season19_20['lhTMratio'])))
season19_20['predD'] = 1/(1+np.exp(-(ol.coef_[2]-ol.coef_[0]*season19_20['lhTMratio']))) - season19_20['predA']
season19_20['predH'] = 1 - season19_20['predA'] - season19_20['predD']

pd.set_option('display.max_columns', 50)
season19_20

date                 Home team                 away team  \
0    2019-10-08 00:00:00           AFC Bournemouth          Sheffield United   
1    2019-10-08 00:00:00                   Burnley               Southampton   
2    2019-10-08 00:00:00            Crystal Palace                   Everton   
3    2019-10-08 00:00:00         Tottenham Hotspur               Aston Villa   
4    2019-10-08 00:00:00                   Watford  Brighton and Hove Albion   
5    2019-10-08 00:00:00           West Ham United           Manchester City   
6    2019-11-08 00:00:00            Leicester City   Wolverhampton Wanderers   
7    2019-11-08 00:00:00         Manchester United                   Chelsea   
8    2019-11-08 00:00:00          Newcastle United                   Arsenal   
9             17/08/2019                   Arsenal                   Burnley   
10            17/08/2019               Aston Villa           AFC Bournemouth   
11            17/08/2019  Brighton and Hove Albion           West Ham United   
12            17/08/2019                   Everton                   Watford   
13            17/08/2019           Manchester City         Tottenham Hotspur   
14            17/08/2019              Norwich City          Newcastle United   
15            17/08/2019               Southampton                 Liverpool   
16            18/08/2019                   Chelsea            Leicester City   
17            18/08/2019          Sheffield United            Crystal Palace   
18            19/08/2019   Wolverhampton Wanderers         Manchester United   
19            23/08/2019               Aston Villa                   Everton   
20            24/08/2019  Brighton and Hove Albion               Southampton   
21            24/08/2019                 Liverpool                   Arsenal   
22            24/08/2019         Manchester United            Crystal Palace   
23            24/08/2019              Norwich City                   Chelsea   
24            24/08/2019          Sheffield United            Leicester City   
25            24/08/2019                   Watford           West Ham United   
26            25/08/2019           AFC Bournemouth           Manchester City   
27            25/08/2019         Tottenham Hotspur          Newcastle United   
28            25/08/2019   Wolverhampton Wanderers                   Burnley   
29            31/08/2019                   Burnley                 Liverpool   
30            31/08/2019                   Chelsea          Sheffield United   
31            31/08/2019            Crystal Palace               Aston Villa   
32            31/08/2019            Leicester City           AFC Bournemouth   
33            31/08/2019           Manchester City  Brighton and Hove Albion   
34            31/08/2019          Newcastle United                   Watford   
35            31/08/2019               Southampton         Manchester United   
36            31/08/2019           West Ham United              Norwich City   
37   2019-01-09 00:00:00                   Arsenal         Tottenham Hotspur   
38   2019-01-09 00:00:00                   Everton   Wolverhampton Wanderers   
39            14/09/2019  Brighton and Hove Albion                   Burnley   
40            14/09/2019                 Liverpool          Newcastle United   
41            14/09/2019         Manchester United            Leicester City   
42            14/09/2019              Norwich City           Manchester City   
43            14/09/2019          Sheffield United               Southampton   
44            14/09/2019         Tottenham Hotspur            Crystal Palace   
45            14/09/2019   Wolverhampton Wanderers                   Chelsea   
46            15/09/2019           AFC Bournemouth                   Everton   
47            15/09/2019                   Watford                   Arsenal   
48            16/09/2019               Aston Villa           West Ham United   
49            20/09/

To identify the most likely outcome, we first identify the largest value in the three columns, predA, predD and predH, which we call Maxprob, and then we create a value for the model's prediction (logitpred) were for the value of Maxprob.

In [14]:
# Result prediction

season19_20['Maxprob'] =season19_20[['predA','predD','predH']].max(axis=1)
season19_20['logitpred']=np.where(season19_20['Maxprob']==season19_20['predA'],'A',\
                               np.where(season19_20['Maxprob']==season19_20['predD'],'D','H'))
season19_20['logittrue']= np.where(season19_20['logitpred'] == season19_20['FTR'],1,0)
season19_20

date                 Home team                 away team  \
0    2019-10-08 00:00:00           AFC Bournemouth          Sheffield United   
1    2019-10-08 00:00:00                   Burnley               Southampton   
2    2019-10-08 00:00:00            Crystal Palace                   Everton   
3    2019-10-08 00:00:00         Tottenham Hotspur               Aston Villa   
4    2019-10-08 00:00:00                   Watford  Brighton and Hove Albion   
5    2019-10-08 00:00:00           West Ham United           Manchester City   
6    2019-11-08 00:00:00            Leicester City   Wolverhampton Wanderers   
7    2019-11-08 00:00:00         Manchester United                   Chelsea   
8    2019-11-08 00:00:00          Newcastle United                   Arsenal   
9             17/08/2019                   Arsenal                   Burnley   
10            17/08/2019               Aston Villa           AFC Bournemouth   
11            17/08/2019  Brighton and Hove Albion           West Ham United   
12            17/08/2019                   Everton                   Watford   
13            17/08/2019           Manchester City         Tottenham Hotspur   
14            17/08/2019              Norwich City          Newcastle United   
15            17/08/2019               Southampton                 Liverpool   
16            18/08/2019                   Chelsea            Leicester City   
17            18/08/2019          Sheffield United            Crystal Palace   
18            19/08/2019   Wolverhampton Wanderers         Manchester United   
19            23/08/2019               Aston Villa                   Everton   
20            24/08/2019  Brighton and Hove Albion               Southampton   
21            24/08/2019                 Liverpool                   Arsenal   
22            24/08/2019         Manchester United            Crystal Palace   
23            24/08/2019              Norwich City                   Chelsea   
24            24/08/2019          Sheffield United            Leicester City   
25            24/08/2019                   Watford           West Ham United   
26            25/08/2019           AFC Bournemouth           Manchester City   
27            25/08/2019         Tottenham Hotspur          Newcastle United   
28            25/08/2019   Wolverhampton Wanderers                   Burnley   
29            31/08/2019                   Burnley                 Liverpool   
30            31/08/2019                   Chelsea          Sheffield United   
31            31/08/2019            Crystal Palace               Aston Villa   
32            31/08/2019            Leicester City           AFC Bournemouth   
33            31/08/2019           Manchester City  Brighton and Hove Albion   
34            31/08/2019          Newcastle United                   Watford   
35            31/08/2019               Southampton         Manchester United   
36            31/08/2019           West Ham United              Norwich City   
37   2019-01-09 00:00:00                   Arsenal         Tottenham Hotspur   
38   2019-01-09 00:00:00                   Everton   Wolverhampton Wanderers   
39            14/09/2019  Brighton and Hove Albion                   Burnley   
40            14/09/2019                 Liverpool          Newcastle United   
41            14/09/2019         Manchester United            Leicester City   
42            14/09/2019              Norwich City           Manchester City   
43            14/09/2019          Sheffield United               Southampton   
44            14/09/2019         Tottenham Hotspur            Crystal Palace   
45            14/09/2019   Wolverhampton Wanderers                   Chelsea   
46            15/09/2019           AFC Bournemouth                   Everton   
47            15/09/2019                   Watford                   Arsenal   
48            16/09/2019               Aston Villa           West Ham United   
49            20/09/

## Evaluating our forecasts

Having generated our forecasts, we want to compare their reliability compared to the betting odds and the forecasts of Nate Silver's 538. We first define a df which consists of those games played in 2020 only (excluding the games from the calendar year 2019 and also the games which have not been played). These are rows 198 to 287 in the data (noting that the first row of data is labeled row 0 in Python). To define a subset of rows 198 to 287 in the data, you write [198:288] - which means that the second row number is not actually included in the subset:

In [15]:
trunc19_20 = season19_20[198:288].copy()
trunc19_20.describe()

notplayed      month        day    year       FTHG       FTAG  \
count       90.0  90.000000  90.000000    90.0  90.000000  90.000000   
mean         0.0   1.744444  13.111111  2020.0   1.566667   0.988889   
std          0.0   0.966415   9.046005     0.0   1.218306   1.075722   
min          0.0   1.000000   1.000000  2020.0   0.000000   0.000000   
25%          0.0   1.000000   7.000000  2020.0   1.000000   0.000000   
50%          0.0   2.000000  12.000000  2020.0   1.000000   1.000000   
75%          0.0   2.000000  21.000000  2020.0   2.000000   1.750000   
max          0.0   8.000000  29.000000  2020.0   4.000000   6.000000   

           B365H      B365D      B365A       TMhome       TMaway     538hpr  \
count  90.000000  90.000000  90.000000    90.000000    90.000000  90.000000   
mean    2.804444   4.314667   4.755890   403.882889   405.895000   0.466000   
std     2.054971   1.497823   3.844079   295.648269   312.390011   0.189741   
min     1.100000   3.100000   1.160000    62.330000    62.330000   0.100000   
25%     1.610000   3.425000   2.390000   207.500000   187.617500   0.350000   
50%     2.100000   3.800000   3.500000   281.700000   279.340000   0.460000   
75%     2.975000   4.475000   5.750000   626.067500   626.067500   0.602500   
max    13.000000  11.000000  19.000000  1140.000000  1140.000000   0.880000   

          538dpr     538apr  lhTMratio   winvalue      predA      predD  \
count  90.000000  90.000000  90.000000  90.000000  90.000000  90.000000   
mean    0.229778   0.304667   0.009730   1.244444   0.329074   0.243114   
std     0.047593   0.171111   1.118824   0.825317   0.128325   0.027735   
min     0.090000   0.030000  -2.906341   0.000000   0.092307   0.139811   
25%     0.210000   0.162500  -0.744887   1.000000   0.232444   0.234920   
50%     0.250000   0.290000   0.024577   1.000000   0.314014   0.253575   
75%     0.260000   0.397500   0.768672   2.000000   0.412362   0.262633   
max     0.300000   0.750000   2.733636   2.000000   0.699727   0.265804   

           predH    Maxprob  logittrue  
count  90.000000  90.000000  90.000000  
mean    0.427812   0.495478   0.466667  
std     0.140547   0.094113   0.501683  
min     0.126155   0.369142   0.000000  
25%     0.324065   0.421253   0.000000  
50%     0.423634   0.476793   0.000000  
75%     0.526293   0.547577   1.000000  
max     0.767882   0.767882   1.000000

As we did with our within-sample forecasts, we will compare both the success rate in predicting the outcome, and the Brier Score.

First the success rate of the forecasts:

In [16]:
# Model success rate
trunc19_20['logittrue'].mean()

0.4666666666666667

A success rate of 46.7% is slightly below the 50% success rate we found with the within-sample predictions, but our main interest is the contrast with the betting odds. We need to identify successes and failures for the bookmaker odds:

In [17]:
trunc19_20['B365true']= np.where(trunc19_20['B365res'] == trunc19_20['FTR'],1,0)
trunc19_20

date                 Home team                 away team  \
198  2020-01-01 00:00:00                   Arsenal         Manchester United   
199  2020-01-01 00:00:00  Brighton and Hove Albion                   Chelsea   
200  2020-01-01 00:00:00                   Burnley               Aston Villa   
201  2020-01-01 00:00:00           Manchester City                   Everton   
202  2020-01-01 00:00:00          Newcastle United            Leicester City   
203  2020-01-01 00:00:00              Norwich City            Crystal Palace   
204  2020-01-01 00:00:00               Southampton         Tottenham Hotspur   
205  2020-01-01 00:00:00                   Watford   Wolverhampton Wanderers   
206  2020-01-01 00:00:00           West Ham United           AFC Bournemouth   
207  2020-02-01 00:00:00                 Liverpool          Sheffield United   
208  2020-10-01 00:00:00          Sheffield United           West Ham United   
209  2020-11-01 00:00:00                   Chelsea                   Burnley   
210  2020-11-01 00:00:00            Crystal Palace                   Arsenal   
211  2020-11-01 00:00:00                   Everton  Brighton and Hove Albion   
212  2020-11-01 00:00:00            Leicester City               Southampton   
213  2020-11-01 00:00:00         Manchester United              Norwich City   
214  2020-11-01 00:00:00         Tottenham Hotspur                 Liverpool   
215  2020-11-01 00:00:00   Wolverhampton Wanderers          Newcastle United   
216  2020-12-01 00:00:00           AFC Bournemouth                   Watford   
217  2020-12-01 00:00:00               Aston Villa           Manchester City   
218           18/01/2020                   Arsenal          Sheffield United   
219           18/01/2020  Brighton and Hove Albion               Aston Villa   
220           18/01/2020           Manchester City            Crystal Palace   
221           18/01/2020          Newcastle United                   Chelsea   
222           18/01/2020              Norwich City           AFC Bournemouth   
223           18/01/2020               Southampton   Wolverhampton Wanderers   
224           18/01/2020                   Watford         Tottenham Hotspur   
225           18/01/2020           West Ham United                   Everton   
226           19/01/2020                   Burnley            Leicester City   
227           19/01/2020                 Liverpool         Manchester United   
228           21/01/2020           AFC Bournemouth  Brighton and Hove Albion   
229           21/01/2020               Aston Villa                   Watford   
230           21/01/2020                   Chelsea                   Arsenal   
231           21/01/2020            Crystal Palace               Southampton   
232           21/01/2020                   Everton          Newcastle United   
233           21/01/2020          Sheffield United           Manchester City   
234           22/01/2020            Leicester City           West Ham United   
235           22/01/2020         Manchester United                   Burnley   
236           22/01/2020         Tottenham Hotspur              Norwich City   
237           23/01/2020   Wolverhampton Wanderers                 Liverpool   
238           29/01/2020           West Ham United                 Liverpool   
239  2020-01-02 00:00:00           AFC Bournemouth               Aston Villa   
240  2020-01-02 00:00:00            Crystal Palace          Sheffield United   
241  2020-01-02 00:00:00            Leicester City                   Chelsea   
242  2020-01-02 00:00:00                 Liverpool               Southampton   
243  2020-01-02 00:00:00         Manchester United   Wolverhampton Wanderers   
244  2020-01-02 00:00:00          Newcastle United              Norwich City   
245  2020-01-02 00:00:00                   Watford                   Everton   
246  2020-01-02 00:00:00           West Ham United  Brighton and Hove Albion   
247  2020-02-02 00:0

Now we can calculate the bookmaker success rate:

In [18]:
trunc19_20['B365true'].mean()

0.4888888888888889

So the bookmaker forecasts are slightly better than our model forecasts - the difference is two games out of 90 correctly forecast. 

Now let's compare the Brier Scores. First, we define a variable for each possible outcome, with the value 1 if this was the actual outcome and zero otherwise:

In [19]:
# Outcome value for calculating Brier Score

trunc19_20['Houtcome']= np.where(trunc19_20['FTR']=='H',1,0)
trunc19_20['Doutcome']= np.where(trunc19_20['FTR']=='D',1,0)
trunc19_20['Aoutcome']= np.where(trunc19_20['FTR']=='A',1,0)
trunc19_20

date                 Home team                 away team  \
198  2020-01-01 00:00:00                   Arsenal         Manchester United   
199  2020-01-01 00:00:00  Brighton and Hove Albion                   Chelsea   
200  2020-01-01 00:00:00                   Burnley               Aston Villa   
201  2020-01-01 00:00:00           Manchester City                   Everton   
202  2020-01-01 00:00:00          Newcastle United            Leicester City   
203  2020-01-01 00:00:00              Norwich City            Crystal Palace   
204  2020-01-01 00:00:00               Southampton         Tottenham Hotspur   
205  2020-01-01 00:00:00                   Watford   Wolverhampton Wanderers   
206  2020-01-01 00:00:00           West Ham United           AFC Bournemouth   
207  2020-02-01 00:00:00                 Liverpool          Sheffield United   
208  2020-10-01 00:00:00          Sheffield United           West Ham United   
209  2020-11-01 00:00:00                   Chelsea                   Burnley   
210  2020-11-01 00:00:00            Crystal Palace                   Arsenal   
211  2020-11-01 00:00:00                   Everton  Brighton and Hove Albion   
212  2020-11-01 00:00:00            Leicester City               Southampton   
213  2020-11-01 00:00:00         Manchester United              Norwich City   
214  2020-11-01 00:00:00         Tottenham Hotspur                 Liverpool   
215  2020-11-01 00:00:00   Wolverhampton Wanderers          Newcastle United   
216  2020-12-01 00:00:00           AFC Bournemouth                   Watford   
217  2020-12-01 00:00:00               Aston Villa           Manchester City   
218           18/01/2020                   Arsenal          Sheffield United   
219           18/01/2020  Brighton and Hove Albion               Aston Villa   
220           18/01/2020           Manchester City            Crystal Palace   
221           18/01/2020          Newcastle United                   Chelsea   
222           18/01/2020              Norwich City           AFC Bournemouth   
223           18/01/2020               Southampton   Wolverhampton Wanderers   
224           18/01/2020                   Watford         Tottenham Hotspur   
225           18/01/2020           West Ham United                   Everton   
226           19/01/2020                   Burnley            Leicester City   
227           19/01/2020                 Liverpool         Manchester United   
228           21/01/2020           AFC Bournemouth  Brighton and Hove Albion   
229           21/01/2020               Aston Villa                   Watford   
230           21/01/2020                   Chelsea                   Arsenal   
231           21/01/2020            Crystal Palace               Southampton   
232           21/01/2020                   Everton          Newcastle United   
233           21/01/2020          Sheffield United           Manchester City   
234           22/01/2020            Leicester City           West Ham United   
235           22/01/2020         Manchester United                   Burnley   
236           22/01/2020         Tottenham Hotspur              Norwich City   
237           23/01/2020   Wolverhampton Wanderers                 Liverpool   
238           29/01/2020           West Ham United                 Liverpool   
239  2020-01-02 00:00:00           AFC Bournemouth               Aston Villa   
240  2020-01-02 00:00:00            Crystal Palace          Sheffield United   
241  2020-01-02 00:00:00            Leicester City                   Chelsea   
242  2020-01-02 00:00:00                 Liverpool               Southampton   
243  2020-01-02 00:00:00         Manchester United   Wolverhampton Wanderers   
244  2020-01-02 00:00:00          Newcastle United              Norwich City   
245  2020-01-02 00:00:00                   Watford                   Everton   
246  2020-01-02 00:00:00           West Ham United  Brighton and Hove Albion   
247  2020-02-02 00:0

Now we derive the bookmaker probabilities from the betting odds. The outcome probability equals 1/(decimal odds). However, if you make this calculation and sum the three possibilities the total is greater than one. This is called the 'overround', or the 'vig' - and represents the profit of the bookmaker. To calculate the implied probability from the betting odds you have to divide by the sum of the three numbers, so that your final probabilities add up to 1 (100%).

We calculate these probabilities below:

In [20]:
trunc19_20['B365HPr']= 1/(trunc19_20['B365H'])/(1/(trunc19_20['B365H'])+ 1/(trunc19_20['B365D'])+ 1/(trunc19_20['B365A']))
trunc19_20['B365DPr']= 1/(trunc19_20['B365D'])/(1/(trunc19_20['B365H'])+ 1/(trunc19_20['B365D'])+ 1/(trunc19_20['B365A']))
trunc19_20['B365APr']= 1/(trunc19_20['B365A'])/(1/(trunc19_20['B365H'])+ 1/(trunc19_20['B365D'])+ 1/(trunc19_20['B365A']))
trunc19_20

date                 Home team                 away team  \
198  2020-01-01 00:00:00                   Arsenal         Manchester United   
199  2020-01-01 00:00:00  Brighton and Hove Albion                   Chelsea   
200  2020-01-01 00:00:00                   Burnley               Aston Villa   
201  2020-01-01 00:00:00           Manchester City                   Everton   
202  2020-01-01 00:00:00          Newcastle United            Leicester City   
203  2020-01-01 00:00:00              Norwich City            Crystal Palace   
204  2020-01-01 00:00:00               Southampton         Tottenham Hotspur   
205  2020-01-01 00:00:00                   Watford   Wolverhampton Wanderers   
206  2020-01-01 00:00:00           West Ham United           AFC Bournemouth   
207  2020-02-01 00:00:00                 Liverpool          Sheffield United   
208  2020-10-01 00:00:00          Sheffield United           West Ham United   
209  2020-11-01 00:00:00                   Chelsea                   Burnley   
210  2020-11-01 00:00:00            Crystal Palace                   Arsenal   
211  2020-11-01 00:00:00                   Everton  Brighton and Hove Albion   
212  2020-11-01 00:00:00            Leicester City               Southampton   
213  2020-11-01 00:00:00         Manchester United              Norwich City   
214  2020-11-01 00:00:00         Tottenham Hotspur                 Liverpool   
215  2020-11-01 00:00:00   Wolverhampton Wanderers          Newcastle United   
216  2020-12-01 00:00:00           AFC Bournemouth                   Watford   
217  2020-12-01 00:00:00               Aston Villa           Manchester City   
218           18/01/2020                   Arsenal          Sheffield United   
219           18/01/2020  Brighton and Hove Albion               Aston Villa   
220           18/01/2020           Manchester City            Crystal Palace   
221           18/01/2020          Newcastle United                   Chelsea   
222           18/01/2020              Norwich City           AFC Bournemouth   
223           18/01/2020               Southampton   Wolverhampton Wanderers   
224           18/01/2020                   Watford         Tottenham Hotspur   
225           18/01/2020           West Ham United                   Everton   
226           19/01/2020                   Burnley            Leicester City   
227           19/01/2020                 Liverpool         Manchester United   
228           21/01/2020           AFC Bournemouth  Brighton and Hove Albion   
229           21/01/2020               Aston Villa                   Watford   
230           21/01/2020                   Chelsea                   Arsenal   
231           21/01/2020            Crystal Palace               Southampton   
232           21/01/2020                   Everton          Newcastle United   
233           21/01/2020          Sheffield United           Manchester City   
234           22/01/2020            Leicester City           West Ham United   
235           22/01/2020         Manchester United                   Burnley   
236           22/01/2020         Tottenham Hotspur              Norwich City   
237           23/01/2020   Wolverhampton Wanderers                 Liverpool   
238           29/01/2020           West Ham United                 Liverpool   
239  2020-01-02 00:00:00           AFC Bournemouth               Aston Villa   
240  2020-01-02 00:00:00            Crystal Palace          Sheffield United   
241  2020-01-02 00:00:00            Leicester City                   Chelsea   
242  2020-01-02 00:00:00                 Liverpool               Southampton   
243  2020-01-02 00:00:00         Manchester United   Wolverhampton Wanderers   
244  2020-01-02 00:00:00          Newcastle United              Norwich City   
245  2020-01-02 00:00:00                   Watford                   Everton   
246  2020-01-02 00:00:00           West Ham United  Brighton and Hove Albion   
247  2020-02-02 00:0

Now we can calculate and compare Brier Scores:

In [21]:
# Model Brier score

Brierlogit = ((trunc19_20['predH'] - trunc19_20['Houtcome'])**2 +(trunc19_20['predD'] - trunc19_20['Doutcome'])**2 +\
             (trunc19_20['predA'] - trunc19_20['Aoutcome'])**2).sum()/89
Brierlogit

0.6199325184748382

In [22]:
# Bookie Brier score (lower is better)

BrierB365 = ((trunc19_20['B365HPr'] - trunc19_20['Houtcome'])**2 +(trunc19_20['B365DPr'] - trunc19_20['Doutcome'])**2 +\
             (trunc19_20['B365APr'] - trunc19_20['Aoutcome'])**2).sum()/89
BrierB365

0.5872181372691697

Once again, we find that the bookmaker Brier Score is slightly lower (better) than the Brier Score of our model, but the gap is even smaller than we found using the within-sample model.

A useful comparison is to generate the crosstabs against the actual result of the bookmaker forecasts and the model forecasts:

In [23]:
pd.crosstab(trunc19_20['FTR'], trunc19_20['B365res'],dropna= True)

B365res   A   H
FTR            
A        11  11
D         7  17
H        11  33

In [24]:
pd.crosstab(trunc19_20['FTR'], trunc19_20['logitpred'],dropna= True)

logitpred   A   H
FTR              
A           9  13
D          10  14
H          11  33

First, note that there were 24 draws out of the 90 games played, and neither bookmaker nor model forecast any draws as the most likely outcome. It seems therefore pointless to compare the middle row of each crosstab. 

Looking at the first row of each crosstab, the bookmaker (B365res) correctly identified 50% of the away wins (11 of the 22), while our model fared less well, getting only 9 of these 22 cases right). 

Looking at the third row of each crosstab, the bookmaker (B365res) correctly identified 75% of the home wins (33 of the 44), and our model had exactly the same success rate.

So we can see that both bookmaker and model are good at forecasting home wins, so-so when it comes to away wins, and useless for predicting draws.

**FiveThirtyEight** is a website devoted to statistical analysis, mostly of political and sports data. It was founded by Nate Silver, who built his reputation on accurate forecasting of US election results based on published opinion poll data (538 is the number of electoral college votes in the US Presidential election- the winner of the election is the candidate who can obtain a majority of these votes, not the popular vote). He also wrote a hugely successful book, The Signal and the Noise, which explores practical applications of statistical prediction, and why humans are notoriously bad at interpreting data to make forecasts (the subtitle of the book is "Why so many predictions fail").

The FiveThirtyEight "Club Soccer Predictions" started in 2017 and now generates game by game predictions for 36 leagues, in the same format that we have followed here- the probability of a home win, draw, and an away win. These probabilities are typically published a week or two ahead of the games being played. The website contains an explanation of how these predictions are generated: https://fivethirtyeight.com/methodology/how-our-club-soccer-predictions-work/. 

The FiveThirtyEight model is a good deal more complicated than our very simple model - the references in the article cite Harmonic means, Massey’s method, the Monte Carlo method, the Poisson process and Ranked probability scores. However, they also rely on TransferMarkt data as a measure of team quality. However, the exact details of the model are not published in a way that would allow anyone to replicate their results. The replicability problem is widespread in the world of statistical analysis, which often makes it difficult to judge the reliability of the model. Here, at least, we have the published forecasts, and so we test the reliability of the model results.

So let's examine how the FiveThirtyEight model compares with the bookmakers and our simple model. We have the FiveThirtyEight outcome probabilities in our dfs, so we just need to generate the most likely outcome and the Brier Score from the data.

In [25]:
# Most likely outcome

trunc19_20['538res']= np.where((trunc19_20['538hpr']>trunc19_20['538dpr']) & (trunc19_20['538hpr']>trunc19_20['538apr']),'H',\
                            np.where((trunc19_20['538dpr']>trunc19_20['538hpr']) & (trunc19_20['538dpr']>trunc19_20['538apr']),'D',\
                                    np.where((trunc19_20['538apr']>trunc19_20['538hpr']) & (trunc19_20['538apr']>trunc19_20['538dpr']),'A',"")))
trunc19_20

date                 Home team                 away team  \
198  2020-01-01 00:00:00                   Arsenal         Manchester United   
199  2020-01-01 00:00:00  Brighton and Hove Albion                   Chelsea   
200  2020-01-01 00:00:00                   Burnley               Aston Villa   
201  2020-01-01 00:00:00           Manchester City                   Everton   
202  2020-01-01 00:00:00          Newcastle United            Leicester City   
203  2020-01-01 00:00:00              Norwich City            Crystal Palace   
204  2020-01-01 00:00:00               Southampton         Tottenham Hotspur   
205  2020-01-01 00:00:00                   Watford   Wolverhampton Wanderers   
206  2020-01-01 00:00:00           West Ham United           AFC Bournemouth   
207  2020-02-01 00:00:00                 Liverpool          Sheffield United   
208  2020-10-01 00:00:00          Sheffield United           West Ham United   
209  2020-11-01 00:00:00                   Chelsea                   Burnley   
210  2020-11-01 00:00:00            Crystal Palace                   Arsenal   
211  2020-11-01 00:00:00                   Everton  Brighton and Hove Albion   
212  2020-11-01 00:00:00            Leicester City               Southampton   
213  2020-11-01 00:00:00         Manchester United              Norwich City   
214  2020-11-01 00:00:00         Tottenham Hotspur                 Liverpool   
215  2020-11-01 00:00:00   Wolverhampton Wanderers          Newcastle United   
216  2020-12-01 00:00:00           AFC Bournemouth                   Watford   
217  2020-12-01 00:00:00               Aston Villa           Manchester City   
218           18/01/2020                   Arsenal          Sheffield United   
219           18/01/2020  Brighton and Hove Albion               Aston Villa   
220           18/01/2020           Manchester City            Crystal Palace   
221           18/01/2020          Newcastle United                   Chelsea   
222           18/01/2020              Norwich City           AFC Bournemouth   
223           18/01/2020               Southampton   Wolverhampton Wanderers   
224           18/01/2020                   Watford         Tottenham Hotspur   
225           18/01/2020           West Ham United                   Everton   
226           19/01/2020                   Burnley            Leicester City   
227           19/01/2020                 Liverpool         Manchester United   
228           21/01/2020           AFC Bournemouth  Brighton and Hove Albion   
229           21/01/2020               Aston Villa                   Watford   
230           21/01/2020                   Chelsea                   Arsenal   
231           21/01/2020            Crystal Palace               Southampton   
232           21/01/2020                   Everton          Newcastle United   
233           21/01/2020          Sheffield United           Manchester City   
234           22/01/2020            Leicester City           West Ham United   
235           22/01/2020         Manchester United                   Burnley   
236           22/01/2020         Tottenham Hotspur              Norwich City   
237           23/01/2020   Wolverhampton Wanderers                 Liverpool   
238           29/01/2020           West Ham United                 Liverpool   
239  2020-01-02 00:00:00           AFC Bournemouth               Aston Villa   
240  2020-01-02 00:00:00            Crystal Palace          Sheffield United   
241  2020-01-02 00:00:00            Leicester City                   Chelsea   
242  2020-01-02 00:00:00                 Liverpool               Southampton   
243  2020-01-02 00:00:00         Manchester United   Wolverhampton Wanderers   
244  2020-01-02 00:00:00          Newcastle United              Norwich City   
245  2020-01-02 00:00:00                   Watford                   Everton   
246  2020-01-02 00:00:00           West Ham United  Brighton and Hove Albion   
247  2020-02-02 00:0

In [26]:
# now identify the forecast successes of FiveThirtyEight

trunc19_20['538true']= np.where(trunc19_20['538res'] == trunc19_20['FTR'],1,0)
trunc19_20

date                 Home team                 away team  \
198  2020-01-01 00:00:00                   Arsenal         Manchester United   
199  2020-01-01 00:00:00  Brighton and Hove Albion                   Chelsea   
200  2020-01-01 00:00:00                   Burnley               Aston Villa   
201  2020-01-01 00:00:00           Manchester City                   Everton   
202  2020-01-01 00:00:00          Newcastle United            Leicester City   
203  2020-01-01 00:00:00              Norwich City            Crystal Palace   
204  2020-01-01 00:00:00               Southampton         Tottenham Hotspur   
205  2020-01-01 00:00:00                   Watford   Wolverhampton Wanderers   
206  2020-01-01 00:00:00           West Ham United           AFC Bournemouth   
207  2020-02-01 00:00:00                 Liverpool          Sheffield United   
208  2020-10-01 00:00:00          Sheffield United           West Ham United   
209  2020-11-01 00:00:00                   Chelsea                   Burnley   
210  2020-11-01 00:00:00            Crystal Palace                   Arsenal   
211  2020-11-01 00:00:00                   Everton  Brighton and Hove Albion   
212  2020-11-01 00:00:00            Leicester City               Southampton   
213  2020-11-01 00:00:00         Manchester United              Norwich City   
214  2020-11-01 00:00:00         Tottenham Hotspur                 Liverpool   
215  2020-11-01 00:00:00   Wolverhampton Wanderers          Newcastle United   
216  2020-12-01 00:00:00           AFC Bournemouth                   Watford   
217  2020-12-01 00:00:00               Aston Villa           Manchester City   
218           18/01/2020                   Arsenal          Sheffield United   
219           18/01/2020  Brighton and Hove Albion               Aston Villa   
220           18/01/2020           Manchester City            Crystal Palace   
221           18/01/2020          Newcastle United                   Chelsea   
222           18/01/2020              Norwich City           AFC Bournemouth   
223           18/01/2020               Southampton   Wolverhampton Wanderers   
224           18/01/2020                   Watford         Tottenham Hotspur   
225           18/01/2020           West Ham United                   Everton   
226           19/01/2020                   Burnley            Leicester City   
227           19/01/2020                 Liverpool         Manchester United   
228           21/01/2020           AFC Bournemouth  Brighton and Hove Albion   
229           21/01/2020               Aston Villa                   Watford   
230           21/01/2020                   Chelsea                   Arsenal   
231           21/01/2020            Crystal Palace               Southampton   
232           21/01/2020                   Everton          Newcastle United   
233           21/01/2020          Sheffield United           Manchester City   
234           22/01/2020            Leicester City           West Ham United   
235           22/01/2020         Manchester United                   Burnley   
236           22/01/2020         Tottenham Hotspur              Norwich City   
237           23/01/2020   Wolverhampton Wanderers                 Liverpool   
238           29/01/2020           West Ham United                 Liverpool   
239  2020-01-02 00:00:00           AFC Bournemouth               Aston Villa   
240  2020-01-02 00:00:00            Crystal Palace          Sheffield United   
241  2020-01-02 00:00:00            Leicester City                   Chelsea   
242  2020-01-02 00:00:00                 Liverpool               Southampton   
243  2020-01-02 00:00:00         Manchester United   Wolverhampton Wanderers   
244  2020-01-02 00:00:00          Newcastle United              Norwich City   
245  2020-01-02 00:00:00                   Watford                   Everton   
246  2020-01-02 00:00:00           West Ham United  Brighton and Hove Albion   
247  2020-02-02 00:0

In [27]:
# 538 model success rate

trunc19_20['538true'].mean()

0.4888888888888889

So FiveThirtyEight had a forecast success rate that was identical to the bookmakers and slightly better than our much simpler model. Now, let's calculate the FiveThirtyEight Brier Score:

In [28]:
# 538 Brier score (lower is better)

Brier538 = ((trunc19_20['538hpr'] - trunc19_20['Houtcome'])**2 +(trunc19_20['538dpr'] - trunc19_20['Doutcome'])**2 +\
             (trunc19_20['538apr'] - trunc19_20['Aoutcome'])**2).sum()/89
Brier538

0.5926

Recall that the Brier Score for the bookmaker was 0.587 and for our simple model it was 0.620. Thus, the FiveThirtyEight Brier Score was marginally worse than the bookmaker's and slightly better than that of our simple model.

Now let's compare all of the crosstabs:

In [29]:
pd.crosstab(trunc19_20['FTR'], trunc19_20['538res'],margins = True,dropna= True)

538res   A   H  All
FTR                
A        9  13   22
D        9  15   24
H        9  35   44
All     27  63   90

In [30]:
pd.crosstab(trunc19_20['FTR'], trunc19_20['B365res'],margins = True,dropna= True)

B365res   A   H  All
FTR                 
A        11  11   22
D         7  17   24
H        11  33   44
All      29  61   90

In [31]:
pd.crosstab(trunc19_20['FTR'], trunc19_20['logitpred'],margins = True,dropna= True)

logitpred   A   H  All
FTR                   
A           9  13   22
D          10  14   24
H          11  33   44
All        30  60   90

Like the bookmaker and our model, FiveThirtyEight never gives the highest probability to a draw.

In terms of away wins, the FiveThirtyEight model's performance is the same as the performance of our model, getting less than half of the results correct (9 out of 22). However, when it comes to home wins, FiveThirtyEight performed better than the bookmaker (63 out of 90).

What is striking, of course, is that all three sets of predictions look rather similar. Indeed, we can confirm this by counting the number of highest probability forecasts that were the same:

In [32]:
# Percentage of highest probability forecasts that were the same for the bookmaker and our model

same_B365_logit = np.where(trunc19_20['B365res'] == trunc19_20['logitpred'],1,0).sum()/90
same_B365_logit

0.8555555555555555

In [33]:
# Percentage of highest probability forecasts that were the same for the bookmaker and the FiveThirtyEight model

same_B365_538 = np.where(trunc19_20['B365res'] == trunc19_20['538res'],1,0).sum()/90
same_B365_538

0.8888888888888888

Overall, the three sources produce very similar forecasts.

## Self Test

The similarities can be confirmed by looking at the correlation between the Brier scores for each set of forecasts. Produce a correlation matrix for the Brier Scores.

In [ ]:
#Your Code Here

## Conclusions

In this notebook, we have moved from within-sample modeling to out-of-sample modeling which can be used to generate forecasts. We found that our simple model based only on the TM value ratio and home advantage performed very close to the odds of the bookmakers, and to a much more complex model produced by FiveThirtyEight.

If you didn't want to go to all the trouble of generating the statisical models, an even simpler rule of thumb would work well- the team most likely to win is the team with the higher TM value. The bigger the gap, the more likely it is that higher TM value team wins. This can be offset to some degree by home advantage - the exact extent could be judged by trial and error. And such a method would also have the advantage of giving some guidance on draws, which are more likely when the TM values are close. 

Why do the bookmakers and the models fail to predict draws? One answer is that the models tend to favour outcomes that occur more frequently, and although there are many draws in soccer, it is much less frequent than a home win, and usually less frequent than an away win. One might imagine that a model such as FiveThirtyEight's which actually forecasts goals scored by each team rather than winning and losing, would be more likely to predict draws, but as we have seen, it isn't.

Our simple model described here will extend quite easily to most other football leagues, since the key to prediction is the variation in TM values, which is quite large in most leagues. This reflects inequalities in the purchasing power of teams. Thus, it can also extend quite easily to leagues where inequalities in spending power is also great - this model works fairly well for Major League Baseball, for example. However, it will not work well for the NFL, since its hard salary cap ensures limited variation in spending between teams, and therefore limited ability to predict based on this difference.

Our model also generates forecasts for the games not played (to date). In the next session we will use our model forecasts to estimate how many points each team would have won from the unplayed games, and then produce a league table based on these outcomes, which can be compared to the league table at the time league play was suspended.

To do this we need to save season19_20 df which contains our forecasts for the full season:

In [34]:
season19_20.to_excel("../../Data/Week 3/forecasts19_20.xlsx")